crates/extensions/tedge_mqtt_bridge/src/lib.rs

```diff
@@ -944,4 +986,412 @@ mod tests {
+            let local_events = [inc!(publish(first_msg)), inc!(publish(second_msg))];
+            let cloud_events = [
+                out!(publish(1)),
+                out!(publish(2)),
+                inc!(puback(2)),
+                inc!(puback(1)),
+            ];
+
+            let bridge = Bridge::default()
+                .with_local_events(local_events)
+                .with_cloud_events(cloud_events)
+                .with_c8y_topics()
+                .process_all_events()
+                .await;
+
+            assert_eq!(
+                bridge.local_client.next_action().unwrap(),
+                Action::Ack(second_msg)
+            );
+            assert_eq!(
+                bridge.local_client.next_action().unwrap(),
+                Action::Ack(first_msg)
+            );
+        }
+
+        #[tokio::test]
+        async fn acknowledges_messages_successfully_following_disconnection() {
+            let first_msg = Publish::new("c8y/s/us", QoS::AtLeastOnce, "first payload");
+            let second_msg = Publish::new("c8y/s/us", QoS::AtLeastOnce, "second payload");
+            let local_events = [inc!(publish(first_msg)), inc!(publish(second_msg))];
+            let cloud_events = [
+                out!(publish(1)),
+                // Abruptly disconnect client
+                inc!(network_error),
+                // Republish message after disconnect
+                out!(publish(1)),
+                inc!(puback(1)),
+                // Then check we successfully acknowledge a future message with the same pkid
+                out!(publish(1)),
+                inc!(puback(1)),
+            ];
+
+            let bridge = Bridge::default()
+                .with_local_events(local_events)
+                .with_cloud_events(cloud_events)
+                .with_c8y_topics()
+                .process_all_events()
+                .await;
+
+            assert_eq!(
+                bridge.local_client.next_action().unwrap(),
+                Action::Ack(first_msg)
+            );
+            assert_eq!(
+                bridge.local_client.next_action().unwrap(),
+                Action::Ack(second_msg)
+            );
+        }
+
+        #[tokio::test]
+        async fn ignores_duplicate_acknowledgement_for_same_message() {
+            let first_msg = Publish::new("c8y/s/us", QoS::AtLeastOnce, "first payload");
+            let second_msg = Publish::new("c8y/s/us", QoS::AtLeastOnce, "second payload");
+            let local_events = [inc!(publish(first_msg)), inc!(publish(second_msg))];
+            let cloud_events = [
+                out!(publish(1)),
+                inc!(puback(1)),
+                // Simulate the cloud sending a second acknowledgement;s
+                inc!(puback(1)),
+                // Then publish the second message
+                out!(publish(2)),
+                inc!(puback(2)),
+            ];
```

Comment:
Looking at this example (duplicated ack) and the two previous ones (inverted ack and disconnection before ack),
I think this is where generated sequences of events would help a lot.

- Publishing 2 local messages
- Expecting 2 local acks 
- For any random sequence of  `out!(publish)`, `inc!(puback)` and `inc!(network_error)` cloud side.
- Not really random sequences as there is some logic relating these messages ... but we should be able to implement filter out sequences making no sense.